### PR TITLE
Add support of port speed set via PORT_TABLE

### PR DESCRIPTION
**Pull request by andriymoroz-mlnx**

### CODE REVIEW

File: orchagent/portsorch.cpp [hunk 1]
```diff
@@ -243,6 +243,77 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
...
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    sai_status_t status;
+
+    attr.id = SAI_PORT_ATTR_SPEED;
+    attr.value.u32 = 0;
+
+    status = sai_port_api->get_port_attribute(port_id, 1, &attr);
+    speed = attr.value.u32;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >speed = attr.value.u32 [](start = 4, length = 22)

if statue != SAI_STATUS_SUCCESS, don't assign speed.

File: orchagent/portsorch.cpp [hunk 2]
```diff
@@ -243,6 +243,77 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
...
+    sai_status_t status;
+
+    // validate speed value
+    attr.id = SAI_PORT_ATTR_SUPPORTED_SPEED;
+    attr.value.u32list.count = 0;
+    attr.value.u32list.list = NULL;
+    status = sai_port_api->get_port_attribute(port_id, 1, &attr);
+    if (status == SAI_STATUS_BUFFER_OVERFLOW)
+    {
+        uint32_t speeds[attr.value.u32list.count];
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >[attr.value.u32list.count] [](start = 23, length = 26)

variable as array size? compilation error?

File: orchagent/portsorch.cpp [hunk 3]
```diff
@@ -243,6 +243,77 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
...
+        attr.value.u32list.list = speeds;
+        status = sai_port_api->get_port_attribute(port_id, 1, &attr);
+        if (status == SAI_STATUS_SUCCESS)
+        {
+            int32_t speed_idx = attr.value.u32list.count;
+            while(speed_idx-- > 0) {
+                if(speed == speeds[speed_idx])
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >== [](start = 25, length = 2)

Do you mean "not equal"?

File: orchagent/portsorch.cpp [hunk 4]
```diff
@@ -243,6 +243,77 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
...
+                    break;
+            }
+            if (speed_idx >= 0)
+            {
+                return true;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >true [](start = 23, length = 4)

Do you mean 'false'?

File: orchagent/portsorch.h [hunk 5]
```diff
@@ -85,6 +85,9 @@ class PortsOrch : public Orch, public Subject
 
     bool setPortAdminStatus(sai_object_id_t id, bool up);
     bool setPortMtu(sai_object_id_t id, sai_uint32_t mtu);
+    bool validatePortSpeed(sai_object_id_t port_id, sai_uint32_t speed);
```
Comment (qiluo-msft) [inline orchagent/portsorch.h]: >validatePortSpeed [](start = 9, length = 17)

Add comment for this function?

File: orchagent/portsorch.cpp (hunk 3 above)
Comment (andriymoroz-mlnx) [inline orchagent/portsorch.cpp]: no, we're looking for the speed value in the list of supported. If equal - found and we can break, if not - keep looking

File: orchagent/portsorch.cpp (hunk 4 above)
Comment (andriymoroz-mlnx) [inline orchagent/portsorch.cpp]: loop goes from count(max) to 0. If idx is still >=0 we're still within the list which means the speed is valid. If not - we'll go down to -1

File: orchagent/portsorch.cpp (hunk 1 above)
Comment (andriymoroz-mlnx) [inline orchagent/portsorch.cpp]: agree

File: orchagent/portsorch.cpp (hunk 2 above)
Comment (andriymoroz-mlnx) [inline orchagent/portsorch.cpp]: no compilation error (C99) but not recommended by the community. Will rewrite to std::vector

File: orchagent/portsorch.cpp (hunk 3 above)
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: Suggest to use std::find()

---
In reply to: [127435840](https://github.com/Azure/sonic-swss/pull/248#discussion_r127435840) [](ancestors = 127435840)

File: orchagent/portsorch.cpp [hunk 6]
```diff
@@ -243,6 +243,79 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
     return true;
 }
 
+bool PortsOrch::validatePortSpeed(sai_object_id_t port_id, sai_uint32_t speed)
+{
+    sai_attribute_t attr;
+    sai_status_t status;
+
+    // validate speed value
+    attr.id = SAI_PORT_ATTR_SUPPORTED_SPEED;
```
Comment (qiluo-msft) [inline orchagent/portsorch.cpp]: >SAI_PORT_ATTR_SUPPORTED_SPEED [](start = 14, length = 29)

Not efficient to get the speed array for each calling.

File: orchagent/portsorch.cpp [hunk 7]
```diff
@@ -243,6 +244,80 @@ bool PortsOrch::setPortMtu(sai_object_id_t id, sai_uint32_t mtu)
     return true;
 }
 
+bool PortsOrch::validatePortSpeed(sai_object_id_t port_id, sai_uint32_t speed)
+{
+    sai_attribute_t attr;
+    sai_status_t status;
+
+
```
Comment (stcheng) [inline orchagent/portsorch.cpp]: one more empty line

Comment (andriymoroz-mlnx) [inline orchagent/portsorch.cpp]: Fixed, thanks

**Pull request merged**